File: pipeline/src/normalizer.py
"""Normalize OKX WebSocket messages to a standard event format."""

from __future__ import annotations

import time
from typing import Any

import msgspec

_DEBUG = True


class BookLevel(msgspec.Struct, frozen=True):
    """Single order book level: (price, size, count)."""

    price: float
    size: float
    count: int


class BookPayload(msgspec.Struct, frozen=True):
    """Payload for book_topn events."""

    n: int
    best_bid: float
    best_ask: float
    bids: list[BookLevel]
    asks: list[BookLevel]


class TradePayload(msgspec.Struct, frozen=True):
    """Payload for trade events."""

    price: float
    size: float
    side: str
    trade_id: str | None


class NormalizedEvent(msgspec.Struct, frozen=True):
    """Normalized market data event."""

    exchange: str
    symbol: str
    channel: str
    event_type: str
    ts_exchange_ms: int
    ts_recv_epoch_ms: int
    ts_recv_mono_ns: int
    ts_decoded_mono_ns: int
    ts_proc_mono_ns: int
    payload: BookPayload | TradePayload


def _parse_levels(raw_levels: list[Any]) -> list[BookLevel]:
    levels: list[BookLevel] = []
    for level in raw_levels:
        if not isinstance(level, list) or len(level) < 4:
            continue
        try:
            levels.append(
                BookLevel(
                    price=float(level[0]),
                    size=float(level[1]),
                    count=int(level[3]),
                )
            )
        except (ValueError, TypeError, IndexError):
            continue
    return levels


def normalize_okx(
    ts_recv_epoch_ms: int,
    ts_recv_mono_ns: int,
    ts_decoded_mono_ns: int,
    msg: dict[str, Any],
) -> list[NormalizedEvent]:
    """Normalize an OKX WebSocket message to NormalizedEvent(s)."""
    if msg.get("event") in ("subscribe", "unsubscribe", "error"):
        return []

    arg = msg.get("arg") or {}
    channel = arg.get("channel")
    data = msg.get("data")

    if not channel or not isinstance(data, list) or not data:
        return []

    inst_id = arg.get("instId")
    if not inst_id:
        return []

    events: list[NormalizedEvent] = []

    if channel == "books5":
        d0 = data[0]
        try:
            ts_exchange_ms = int(d0.get("ts", "0"))
        except (ValueError, TypeError):
            return []

        bids = _parse_levels(d0.get("bids") or [])
        asks = _parse_levels(d0.get("asks") or [])

        payload = BookPayload(
            n=5,
            best_bid=(bids[0].price if bids else 0.0),
            best_ask=(asks[0].price if asks else 0.0),
            bids=bids,
            asks=asks,
        )

        ts_proc_mono_ns = time.monotonic_ns()
        if _DEBUG:
            if ts_decoded_mono_ns < ts_recv_mono_ns:
                raise RuntimeError(
                    f"Invariant violated: decoded_ns ({ts_decoded_mono_ns}) < recv_ns ({ts_recv_mono_ns})"
                )
            if ts_proc_mono_ns < ts_decoded_mono_ns:
                raise RuntimeError(
                    f"Invariant violated: proc_ns ({ts_proc_mono_ns}) < decoded_ns ({ts_decoded_mono_ns})"
                )

        events.append(
            NormalizedEvent(
                exchange="okx",
                symbol=inst_id,
                channel="books5",
                event_type="book_topn",
                ts_exchange_ms=ts_exchange_ms,
                ts_recv_epoch_ms=ts_recv_epoch_ms,
                ts_recv_mono_ns=ts_recv_mono_ns,
                ts_decoded_mono_ns=ts_decoded_mono_ns,
                ts_proc_mono_ns=ts_proc_mono_ns,
                payload=payload,
            )
        )

    elif channel == "trades":
        for d in data:
            try:
                ts_exchange_ms = int(d.get("ts", "0"))
            except (ValueError, TypeError):
                continue

            try:
                payload = TradePayload(
                    price=float(d["px"]),
                    size=float(d["sz"]),
                    side=d["side"],
                    trade_id=d.get("tradeId"),
                )
            except (KeyError, ValueError, TypeError):
                continue

            ts_proc_mono_ns = time.monotonic_ns()
            if _DEBUG:
                if ts_decoded_mono_ns < ts_recv_mono_ns:
                    raise RuntimeError(
                        f"Invariant violated: decoded_ns ({ts_decoded_mono_ns}) < recv_ns ({ts_recv_mono_ns})"
                    )
                if ts_proc_mono_ns < ts_decoded_mono_ns:
                    raise RuntimeError(
                        f"Invariant violated: proc_ns ({ts_proc_mono_ns}) < decoded_ns ({ts_decoded_mono_ns})"
                    )

            events.append(
                NormalizedEvent(
                    exchange="okx",
                    symbol=inst_id,
                    channel="trades",
                    event_type="trade",
                    ts_exchange_ms=ts_exchange_ms,
                    ts_recv_epoch_ms=ts_recv_epoch_ms,
                    ts_recv_mono_ns=ts_recv_mono_ns,
                    ts_decoded_mono_ns=ts_decoded_mono_ns,
                    ts_proc_mono_ns=ts_proc_mono_ns,
                    payload=payload,
                )
            )

    return events

File: pipeline/src/normalizer.cpp
#include "normalizer.hpp"

#include "time_helpers.hpp"

namespace quantumflow::pipeline {

namespace {

// OKX sends numeric fields as JSON strings (e.g. "43000.5"). Accept both string
// and number forms; return std::nullopt on failure (mirrors Python's try/except).
std::optional<double> to_double(const nlohmann::json& v) {
    try {
        if (v.is_string()) {
            const std::string& s = v.get_ref<const std::string&>();
            if (s.empty()) return std::nullopt;
            size_t pos = 0;
            double d = std::stod(s, &pos);
            return d;
        }
        if (v.is_number()) {
            return v.get<double>();
        }
    } catch (...) {
    }
    return std::nullopt;
}

std::optional<int64_t> to_int(const nlohmann::json& v) {
    try {
        if (v.is_string()) {
            const std::string& s = v.get_ref<const std::string&>();
            if (s.empty()) return std::nullopt;
            size_t pos = 0;
            // Parse as double then truncate to mirror Python int() on numeric strings,
            // but OKX counts/timestamps are integral strings so stoll is exact.
            long long n = std::stoll(s, &pos);
            return static_cast<int64_t>(n);
        }
        if (v.is_number_integer() || v.is_number_unsigned()) {
            return v.get<int64_t>();
        }
        if (v.is_number_float()) {
            return static_cast<int64_t>(v.get<double>());
        }
    } catch (...) {
    }
    return std::nullopt;
}

// Parse a list of [price, size, _, count] levels. Skips malformed entries.
std::vector<BookLevel> parse_levels(const nlohmann::json& raw_levels) {
    std::vector<BookLevel> levels;
    if (!raw_levels.is_array()) return levels;
    for (const auto& level : raw_levels) {
        if (!level.is_array() || level.size() < 4) continue;
        auto price = to_double(level[0]);
        auto size = to_double(level[1]);
        auto count = to_int(level[3]);
        if (!price || !size || !count) continue;
        levels.push_back(BookLevel{*price, *size, *count});
    }
    return levels;
}

const nlohmann::json* find_member(const nlohmann::json& obj, const char* key) {
    if (!obj.is_object()) return nullptr;
    auto it = obj.find(key);
    if (it == obj.end()) return nullptr;
    return &(*it);
}

} // namespace

std::vector<NormalizedEvent> normalize_okx(
    int64_t ts_recv_epoch_ms,
    int64_t ts_recv_mono_ns,
    int64_t ts_decoded_mono_ns,
    const nlohmann::json& msg) {

    std::vector<NormalizedEvent> events;
    if (!msg.is_object()) return events;

    // Skip subscription acks / errors.
    if (const auto* ev = find_member(msg, "event"); ev && ev->is_string()) {
        const std::string& e = ev->get_ref<const std::string&>();
        if (e == "subscribe" || e == "unsubscribe" || e == "error") return events;
    }

    const auto* arg = find_member(msg, "arg");
    const auto* data = find_member(msg, "data");
    if (!arg || !data || !data->is_array() || data->empty()) return events;

    const auto* channel_j = find_member(*arg, "channel");
    const auto* inst_j = find_member(*arg, "instId");
    if (!channel_j || !channel_j->is_string()) return events;
    if (!inst_j || !inst_j->is_string()) return events;

    const std::string channel = channel_j->get<std::string>();
    const std::string inst_id = inst_j->get<std::string>();
    if (channel.empty() || inst_id.empty()) return events;

    if (channel == "books5") {
        const auto& d0 = (*data)[0];
        const auto* ts_j = find_member(d0, "ts");
        int64_t ts_exchange_ms = 0;
        if (ts_j) {
            auto ts = to_int(*ts_j);
            if (!ts) return events;
            ts_exchange_ms = *ts;
        }

        BookPayload payload;
        payload.n = 5;
        if (const auto* bids = find_member(d0, "bids")) payload.bids = parse_levels(*bids);
        if (const auto* asks = find_member(d0, "asks")) payload.asks = parse_levels(*asks);
        payload.best_bid = payload.bids.empty() ? 0.0 : payload.bids[0].price;
        payload.best_ask = payload.asks.empty() ? 0.0 : payload.asks[0].price;

        NormalizedEvent ev;
        ev.exchange = "okx";
        ev.symbol = inst_id;
        ev.channel = "books5";
        ev.event_type = "book_topn";
        ev.ts_exchange_ms = ts_exchange_ms;
        ev.ts_recv_epoch_ms = ts_recv_epoch_ms;
        ev.ts_recv_mono_ns = ts_recv_mono_ns;
        ev.ts_decoded_mono_ns = ts_decoded_mono_ns;
        ev.ts_proc_mono_ns = now_mono_ns();
        ev.payload = std::move(payload);
        events.push_back(std::move(ev));

    } else if (channel == "trades") {
        for (const auto& d : *data) {
            const auto* ts_j = find_member(d, "ts");
            if (!ts_j) continue;
            auto ts_exchange_ms = to_int(*ts_j);
            if (!ts_exchange_ms) continue;

            const auto* px = find_member(d, "px");
            const auto* sz = find_member(d, "sz");
            const auto* side = find_member(d, "side");
            if (!px || !sz || !side || !side->is_string()) continue;

            auto price = to_double(*px);
            auto size = to_double(*sz);
            if (!price || !size) continue;

            TradePayload payload;
            payload.price = *price;
            payload.size = *size;
            payload.side = side->get<std::string>();
            if (const auto* tid = find_member(d, "tradeId"); tid && tid->is_string()) {
                payload.trade_id = tid->get<std::string>();
            }

            NormalizedEvent ev;
            ev.exchange = "okx";
            ev.symbol = inst_id;
            ev.channel = "trades";
            ev.event_type = "trade";
            ev.ts_exchange_ms = *ts_exchange_ms;
            ev.ts_recv_epoch_ms = ts_recv_epoch_ms;
            ev.ts_recv_mono_ns = ts_recv_mono_ns;
            ev.ts_decoded_mono_ns = ts_decoded_mono_ns;
            ev.ts_proc_mono_ns = now_mono_ns();
            ev.payload = std::move(payload);
            events.push_back(std::move(ev));
        }
    }

    return events;
}

} // namespace quantumflow::pipeline
